core/src/main/java/com/yahoo/ycsb/TimeseriesDB.java

```diff
@@ -0,0 +1,282 @@
+        TimeSeriesWorkload.PAIR_DELIMITER_PROPERTY_DEFAULT);
+    queryTimeSpanDelimiter = getProperties().getProperty(
+        TimeSeriesWorkload.QUERY_TIMESPAN_DELIMITER_PROPERTY,
+        TimeSeriesWorkload.QUERY_TIMESPAN_DELIMITER_PROPERTY_DEFAULT);
+    deleteDelimiter = getProperties().getProperty(
+        TimeSeriesWorkload.DELETE_DELIMITER_PROPERTY,
+        TimeSeriesWorkload.DELETE_DELIMITER_PROPERTY_DEFAULT);
+    timestampUnit = TimeUnit.valueOf(getProperties().getProperty(
+        TimeSeriesWorkload.TIMESTAMP_UNITS_PROPERTY,
+        TimeSeriesWorkload.TIMESTAMP_UNITS_PROPERTY_DEFAULT));
+    groupByKey = getProperties().getProperty(
+        TimeSeriesWorkload.GROUPBY_KEY_PROPERTY,
+        TimeSeriesWorkload.GROUPBY_KEY_PROPERTY_DEFAULT);
+    downsamplingKey = getProperties().getProperty(
+        TimeSeriesWorkload.DOWNSAMPLING_KEY_PROPERTY,
+        TimeSeriesWorkload.DOWNSAMPLING_KEY_PROPERTY_DEFAULT);
+    downsamplingFunction = TimeseriesDB.AggregationOperation.valueOf(getProperties()
+        .getProperty(TimeSeriesWorkload.DOWNSAMPLING_FUNCTION_PROPERTY, DOWNSAMPLING_FUNCTION_PROPERTY_DEFAULT));
+    downsamplingInterval = Integer.valueOf(getProperties()
+        .getProperty(TimeSeriesWorkload.DOWNSAMPLING_INTERVAL_PROPERTY, DOWNSAMPLING_INTERVAL_PROPERTY_DEFAULT));
+
+    test = Boolean.parseBoolean(getProperties().getProperty(TEST_PROPERTY, TEST_PROPERTY_DEFAULT));
+    debug = Boolean.parseBoolean(getProperties().getProperty(DEBUG_PROPERTY, DEBUG_PROPERTY_DEFAULT));
+  }
+
+  @Override
+  public final Status read(String table, String key, Set<String> fields, Map<String, ByteIterator> result) {
+    Map<String, List<String>> tagQueries = new HashMap<>();
+    Long timestamp = null;
+    for (String field : fields) {
+      if (field.startsWith(timestampKey)) {
+        String[] timestampParts = field.split(tagPairDelimiter);
+        if (timestampParts[1].contains(queryTimeSpanDelimiter)) {
+          // seems like this should be a more elaborate query.
+          // for now we don't support querying ranges
+          // TODO: Support Timestamp range queries
+          return Status.NOT_IMPLEMENTED;
+        }
+        timestamp = Long.valueOf(timestampParts[1]);
+      } else {
+        String[] queryParts = field.split(tagPairDelimiter);
+        tagQueries.computeIfAbsent(queryParts[0], k -> new ArrayList<>()).add(queryParts[1]);
+      }
+    }
+
+    return read(table, timestamp, tagQueries);
+  }
+
+  /**
+   * Read a record from the database. Each value from the result will be stored in a HashMap
+   *
+   * @param metric    The name of the metric
+   * @param timestamp The timestamp of the record to read.
+   * @param tags      actual tags that were want to receive (can be empty)
+   * @return Zero on success, a non-zero error code on error or "not found".
+   */
+  protected abstract Status read(String metric, Long timestamp, Map<String, List<String>> tags);
```

Comment:
Still need to pass through `table` though a lot of TSDs will ignore it, and add the end timestamp and convert `Long` to `long` to use the primitive.

Reply:
`metric` is the equivalent to `table`. See also the callsite (`return read(table, timestamp, tagQueries;`)

As established above, range queries for reads make no sense. Not sure what you mean by "add the end timestamp"